### PR TITLE
[WIP] Adds Departmental Research Trees

### PR DESCRIPTION
**Pull request by TheVekter**
## About The Pull Request
I am STILL very much working on this and trying to get as much help as I can to finish it off.

As of this PR, each department will have its own research tree. Specifically, this refers to:

- Science
- Cargo and Mining
- Engineering
- Medical
- Security
- Service

This means six different point budgets, six different R&D consoles, six different sets of research. This has a few purposes but is basically designed to keep the whole crew invested in what's being researched and how it affects their department instead of just one department (Science) doing all of it.

This isn't going to change that heads can only do R&D now. That change was done to give heads a bit more mechanical agency over things as opposed to just being the RP leaders and rev targets. This seems odd now but once we roll out more changes, it'll make more sense.

This adds an extra step when adding a tech node - you need to assign the variable "departmental_flags" to each new node, then use the bitflag for what department should be able to research it. If you use the ALL flag, everyone can.

There is very little collision in this system, Mining can now just research mining tech for itself. Security can research whatever weapons it wants. The only real overlap is that a few prerequisites that are assigned to either Science or Engineering, but not researching those would severely screw the department in question, so I don't see that being a problem.

Not having a head is still a pretty valid issue to have, but there are a few solutions including having the AI let you in to grab the board, hacking in to get the board, or getting promoted to head access. Each console is in each head's office, so there's only one door you need to hack to get at it.

One question might be as to why the QM has one - the HoP is already over multiple departments and mining's research is important enough to them that it should be handled by someone close by, not the HoP.

## Why It's Good For The Game

This opens up a lot of expansion room to R&D. As opposed to it only really benefiting one department as a majority, it's something everyone gets involved in. The plan is eventually to have separate trees you can choose to go down which should make it more engaging than just "farm up to collect everything in the tree". It also, as mentioned above, removes a pretty large source of friction between departments. Science being lazy? Fuck 'em, research your own crap.

## Changelog
:cl:
add: All research has been split up per department to assist with efficiency aboard the station.
add: Each head of staff and the Quartermaster now start with a research console for their department in their quarters. The core R&D console on the bridge has not been affected.
add: Each department will now accrue points for their own department, as opposed to one generic pool of research points.
/:cl:

### CODE REVIEW

File: code/controllers/subsystem/research.dm [hunk 1]
```diff
@@ -20,33 +21,42 @@ SUBSYSTEM_DEF(research)
 
 	var/list/obj/machinery/rnd/server/servers = list()
 
-	var/list/techweb_nodes_starting = list()	//associative id = TRUE
-	var/list/techweb_categories = list()		//category name = list(node.id = TRUE)
-	var/list/techweb_boost_items = list()		//associative double-layer path = list(id = list(point_type = point_discount))
-	var/list/techweb_nodes_hidden = list()		//Node ids that should be hidden by default.
+	var/list/techweb_nodes_starting = list()		//associative id = TRUE
```
Comment (Qustinnus) [inline code/controllers/subsystem/research.dm:24]: these need to have autodoc comments

File: code/datums/dog_fashion.dm [hunk 2]
```diff
@@ -179,3 +178,7 @@
 	name = "Festive REAL_NAME"
 	desc = "Ready to party!"
 	obj_icon_state = "festive"
+
+/datum/dog_fashion/head/bishopmitre
```
Comment (Qustinnus) [inline code/datums/dog_fashion.dm:182]: wat

File: code/controllers/subsystem/research.dm [hunk 3]
```diff
@@ -12,7 +12,7 @@ SUBSYSTEM_DEF(research)
 	var/datum/techweb/admin/admin_tech
 	var/datum/techweb_node/error_node/error_node	//These two are what you get if a node/design is deleted and somehow still stored in a console.
 	var/datum/design/error_design/error_design
-	var/datum/techweb_node/department			///What department can research this node.
+	var/datum/techweb_node/epartmental_flags		///What department can research this node.
```
Comment (Shadark) [inline code/controllers/subsystem/research.dm:15]: departmental_flags I guess?

File: code/modules/research/departmental_rdconsole.dm [hunk 4]
```diff
@@ -0,0 +1,48 @@
+/obj/machinery/computer/rdconsole/department
+	name = "department research console"
+	desc = "A special research console designed to permit each department to research technology without the assistance of the Science department."
+	icon_screen = "rdcomp"
+	icon_keyboard = "rd_key"
+	allowed_department_flags = DEPARTMENTAL_FLAG_ALL
```
Comment (Qustinnus) [inline code/modules/research/departmental_rdconsole.dm:6]: DEPARTMENTAL_FLAG_ALL was removed

File: code/controllers/subsystem/research.dm (hunk 3 above)
Comment (TheVekter) [reply code/controllers/subsystem/research.dm:15]: Yeah?

**Review by ghost — CHANGES_REQUESTED** [review]
some small notes, a lot of the rest looks fine and doesn't really touch the core of techwebs from a system perspective

File: code/controllers/subsystem/research.dm (hunk 3 above)
Comment (Qustinnus) [reply code/controllers/subsystem/research.dm:15]: it says epartmental flags

File: code/datums/dog_fashion.dm (hunk 2 above)
Comment (TheVekter) [reply code/datums/dog_fashion.dm:182]: That's from an old attempt at getting Ian to wear the bishop hat. I have no clue how that ended up in here but good catch.

File: code/controllers/subsystem/research.dm (hunk 1 above)
Comment (TheVekter) [reply code/controllers/subsystem/research.dm:24]: Gonna go through that hell once I'm done.

Comment (Qustinnus) [reply code/controllers/subsystem/research.dm:24]: and that applies to all the vars u touched  😳

**Comment by ATH1909** [comment]
image

image

image

https://youtu.be/_J6-3l3hCm0

**Comment by TheVekter** [comment]
WOULD YOU LIKE TO CODE IT INSTEAD ATH

**Comment by Qustinnus** [comment]
also dont open wip PRs if theyre not testmergeable not epic

File: code/controllers/subsystem/research.dm (hunk 3 above)
Comment (ExcessiveUseOfCobblestone) [reply code/controllers/subsystem/research.dm:15]: the d comes later 😏

**Comment by TheVekter** [comment]
> 
> 
> also dont open wip PRs if theyre not testmergeable not epic

I mostly did this because I need help finishing the PR and the person who was helping me is now either ignoring me or has died.

**Comment by Qustinnus** [comment]
didnt i join the experisci discord group specifically so u two could ask questions :^)

**Comment by GhostEcho** [comment]
one thing I am mildly concerned about is that by removing that source of friction, you move us towards a bit more of a isolated sterile environment.

SS13 is built around arising conflicts, and having to deal with science being little shits, as much as I hate it as a miner, is something I actually prefer to stay in.encourages more player interaction potentially positive if Science is friendly and dutiful, or negative if they're dicks, maybe get into a fist fight and end up arrested because I had to break in to do it myself.

With this, I'd just ask my head and then right back to work it sounds like. I'm not sure if a system like this really benefits overall.

**Comment by ExcessiveUseOfCobblestone** [comment]
Station cooperation is terribad unless science players dominate in the dynamic, in which case it's a critical aspect of the game that they own the most important station-wide progression system.

**Comment by TheVekter** [comment]
The scope of those interactions is basically either "'Hey, can you research this?' 'Sure.' Five minutes later 'Okay it's done.' 'Thanks!'" or Mining revolting against Science. That last part is more of a problem with how mining is currently set up but in lieu of a full mining rework, I'm happy to offer this.

Experimental research is going to open up a lot more opportunities for cooperation in how experiments are completed and their requirements without the concern of people wanting to bash each others' heads in.

**Comment by TheVekter** [comment]
Also I'm closing this until I can cba to fix these goddamn merges (so probably tonight)

**Comment by ATH1909** [comment]
> one thing I am mildly concerned about is that by removing that source of friction, you move us towards a bit more of a isolated sterile environment.
> 
> SS13 is built around arising conflicts, and having to deal with science being little shits, as much as I hate it as a miner, is something I actually prefer to stay in.encourages more player interaction potentially positive if Science is friendly and dutiful, or negative if they're dicks, maybe get into a fist fight and end up arrested because I had to break in to do it myself.
> 
> With this, I'd just ask my head and then right back to work it sounds like. I'm not sure if a system like this really benefits overall.

That friction will still be there, but it'll be with the heads of staff and/or the silicons instead of Science. Unfortunately, everyone still asks Science to research techs, because when they think "R&D", they (logically) think the Science department, not the Command department. Maybe the solution is to just add departmental techwebs and move R&D back to science (and the bridge and the offices of each of the heads of staff, I suppose)? I'm not really seeing what we're gaining by having the command staff do R&D instead of Science.

**Comment by ExcessiveUseOfCobblestone** [comment]
RND when in the hands of science is geared heavily towards the sci/mining dynamic (rightfully so) so it would be nice if each department could use the mechanic in a tangible way rather than an afterthought once the meta techs have been researched.

Really would like to see the meta around parts / mining rush be shaken up. It's a braindead choice you have to make literally every round.

**Comment by ATH1909** [comment]
> RND when in the hands of science is geared heavily towards the sci/mining dynamic (rightfully so) so it would be nice if each department could use the mechanic in a tangible way rather than an afterthought once the meta techs have been researched.
> 
> Really would like to see the meta around parts / mining rush be shaken up. It's a braindead choice you have to make literally every round.

Still don't see a reason why command should do it instead of sci.

**Comment by TheVekter** [comment]
> Still don't see a reason why command should do it instead of sci. If we're doing departmental trees, then each department will already be "using the mechanic in a tangible way rather than an afterthought", and the "meta around parts / mining rush" will be shaken up anyway.

Command is doing this because I want to give heads more actual authority over what's being done in their department as opposed to them just sort of being de facto leaders because they're assumed to be such. They're also doing it because, once I overhaul tech trees in the coming months, there will be actual tangible choices needed to be made in techs that will affect what the department can or can't do. Maybe Cargo researches a way to just teleport stuff straight to a department instead of having to mule it. Or they could take the express console instead and get stuff instantly from Centcom. Engineering can choose if they want to prioritize on making better gas mixtures to sell or on making it easier to build cool structures. Robotics could go with making big stompy combat mechs or quicker, more nimble ones that are better suited to heavy lifting or mining.

If you give that choice to players, it'd lead to more friction due to one thing being picked over another by a member of staff. If we're going in this direction (and I'd very much like to), it's going to help a lot for us to have a single decision-making role in each department to settle those arguments.

Yes, I know QM isn't a head, hush.

**Comment by ATH1909** [comment]
Why not let both Science AND Command research techs, though? Sometimes your local head of staff will fuck off and be hard to find, so if that happens, you could go to Science instead.

**Pull request closed without merging**